001-clarifications-eval.txt

```diff
@@ -0,0 +1,754 @@
+    2.5.1 DWARF Expression Evaluation Context
+
+    A DWARF expression is evaluated in a context that can include a
+    number of context elements. If multiple context elements are
+    specified then they must be self consistent or the result of the
+    evaluation is undefined. The context elements that can be specified
+    are:
+
+    1. A current result kind
+
+        The kind of result required by the DWARF expression evaluation.
+        If specified it can be a location description or a value.
+
+    2. A current thread
+
+        The target architecture thread identifier of the source program
+        thread of execution for which a user presented expression is
+        currently being evaluated.
+
+        It is required for operations that are related to target
+        architecture threads.
+
+        [non-normative] For example: the DW_OP_regval_type operation.
+
+    3. A current call frame
+
+        The target architecture call frame identifier. It identifies a
+        call frame that corresponds to an active invocation of a
+        subprogram in the current thread. It is identified by its
+        address on the call stack. The address is referred to as the
+        Canonical Frame Address (CFA). The call frame information is
+        used to determine the CFA for the call frames of the current
+        thread’s call stack (see 6.4 Call Frame Information).
+
+        It is required for operations that specify target architecture
+        registers to support virtual unwinding of the call stack.
+
+        [non-normative] For example: the DW_OP_*reg* operations.
+
+        If specified, it must be an active call frame in the current
+        thread. Otherwise the result is undefined.
+
+        If it is the currently executing call frame, then it is termed
+        the top call frame.
+
+    4. A current program location
+
+        The target architecture program location corresponding to the
+        current call frame of the current thread.
+
+        The program location of the top call frame is the target
+        architecture program counter for the current thread. The call
+        frame information is used to obtain the value of the return
+        address register to determine the program location of the other
+        call frames (see 6.4 Call Frame Information).
+
+        It is required for the evaluation of location list expressions
+        to select amongst multiple program location ranges. It is
+        required for operations that specify target architecture
+        registers to support virtual unwinding of the call stack (see
+        6.4 Call Frame Information).
+
+        If specified:
+
+          * If the current call frame is the top call frame, it must be
+            the current target architecture program location.
+
+          * If the current call frame F is not the top call frame, it
+            must be the program location associated with the call site
```

Comment:
Those are good points. The text is treating call site and return address as being the same thing. I am not sure if DWARF defines if CFA should calculate the return address or the call site address. What does gdb do?

Although the two values are very similar, since loclists use the program location to select the expression to evaluate, it can make a difference.